### ISSUE TITLE
Architecture for Text Classification

### ISSUE THREAD

**Issue by bkj** [fchollet/keras#1637]
Does anyone have heuristic architecture guidelines for building a network for binary text classification?

Current, I am using 

```
model = Sequential()
model.add(LSTM(output_dim = 64, input_dim = 64, return_sequences = False))
model.add(Dense(16))
model.add(Dropout(0.5))
model.add(Dense(2))
model.add(Activation('sigmoid'))
model.compile(loss='binary_crossentropy', optimizer='rmsprop')
```

because I saw something similar in various examples.  The input here is a 3d array of pretrained word2vec representations w/ shape (n_docs X words_per_doc X dimension_per_word)

**Comment by wxs** [comment]
Briefly: you'll want your final `Dense` to be of size "1" for a binary classification task, not 2. That gives you a single output, with "0" representing one class, and "1" the other.

As to the details of how large to make each layer: experiment! I suspect you'll want a larger hidden state on your LSTM, and larger word vectors though.

**Comment by bkj** [comment]
OK thanks for the tips.

What does the larger hidden state on the LSTM do though, exactly?
Is an RNN better than a CNN?  I see there are examples of both in this repo.
Should we expect this network to work particularly well?  Or should I be making it significantly deeper?

**Comment by wxs** [comment]
It really depends on your particular classification problem! There's a bit of a black art to figuring out the ideal architecture. RNNs often work well for this kind of problem. There's no one right answer that's true for every problem.

A larger hidden state on the LSTM allows it to "remember" more things over time that may prove relevant in its final classification.

**Comment by pasky** [comment]
When classifying sentences, it seems to me that CNN works better if you have less training data. An interesting comparison is e.g. <a href="http://arxiv.org/pdf/1510.03753.pdf">1510.03753</a>. But I don't know how well will CNN scale for whole documents.  I have a few pet research ideas about this...

On the RNN side, you may also consider a bidirectional LSTM - add another LSTM that go_backward=True, then concat or sum them.  Then, there's tons more ideas - CNN on top of sequence-producing LSTM, trying GRU instead, etc.

On another note, I'd suggest considering a much wider hidden Dense layer than input/4. It may or may not work well. If your data is big, you may want to allow embedding matrix co-learning, use the Embedding layer (you can preinit the embedding matrix by word2vec or glove). Also, didn't you mean your output Dense layer to have output_dim=1?

**Comment by farizrahman4u** [comment]
Conv-RNNs always give me better results than RNNs alone. See my example : https://github.com/fchollet/keras/blob/master/examples/imdb_cnn_lstm.py

**Comment by bkj** [comment]
Great thanks for the tips everyone.

Output_dim = 2 also works, if your output looks like

```
[[0,1], [0, 1], [1, 0], ...]
```

though I agree it's a little sloppy. I'll try a bidirectional LSTM now.  Conv-RNNs also sound very interesting.

**Comment by farizrahman4u** [comment]
For binary classification, you can have 2 outputs and then do a softmax over them or a single output with a sigmoid on top.

Use 2 outputs + softmax if:
* Your classes are discrete. Example: Cats and dogs. (75% cat and 25% dog doesn't make sense. Its either cat or dog.)

Use one output with sigmoid if:
* Your classes are continuous. Example: Binary sentiment analysis.

**Comment by bkj** [comment]
Great thanks for the tip.  Seems to make sense.

**Comment by wxs** [comment]
@farizrahman4u do you have any references for the 2 output softmax being preferred if you have discrete classes? I don't see why it would be different that a single output + sigmoid. The main difference I see is that for the 2-output you have twice the parameters so conceivably could learn more. But for equivalent numbers of parameters (so e.g. half as many hidden nodes in the 2 output case) I don't see why you would expect a difference.

**Comment by tboquet** [comment]
I think @wxs point is correct, you can use either a softmax or a sigmoid. It think it's also possible to show that you wont have better estimators/performances if you use a softmax with twice the number of parameters over a sigmoid (if you take a simple logistic regression problem without all the feature extraction part i.e. the problem is convex). The softmax will also give you this 75% / 25% result.

**Comment by farizrahman4u** [comment]
Here is a script to play around with:

```python
from seq2seq.models import Sequential
from keras.layers import Dense, Activation
import numpy as np
import keras.backend as K

def categorical(x):
	#Function to convert 1s and 0s to [1, 0] and [0, 1]
	y = [[0, 1] if x[j][0] == 0 else [1, 0] for j in range(x.shape[0])]
	y = np.array(y)
	return y

def size(model): # Compute number of params in a model (the actual number of floats)
	return sum([sum(K.get_value(w).shape) for w in model.trainable_weights])

x = np.random.random((100000, 10))

#continuous classes: best results when using model1
y1 = np.array([[1/(1 + np.power(np.e, -np.sum(x[j])))] for j in range(x.shape[0])])
#descrete classes : best results when using model2
y2 = np.array([[int(np.sum(x[j])) % 2] for j in range(x.shape[0])])

#One output, sigmoid
model1 = Sequential()
model1.add(Dense(input_dim=10, output_dim=10))
model1.add(Dense(1))
model1.add(Dense(1))
model1.add(Activation('sigmoid'))

#2 outputs, softmax
model2 = Sequential()
model2.add(Dense(input_dim=10, output_dim=10))
model2.add(Dense(2))
model2.add(Activation('softmax'))


print size(model1), size(model2)

model1.compile(loss='mse', optimizer='sgd')
model1.fit(x, y2, validation_split=0.05, nb_epoch=10)

model2.compile(loss='mse', optimizer='sgd')
model2.fit(x, categorical(y2), validation_split=0.05, nb_epoch=10)
```

**Comment by farizrahman4u** [comment]
Here is a script to play around with:

```python
from seq2seq.models import Sequential
from keras.layers import Dense, Activation
import numpy as np
import keras.backend as K

def categorical(x):
	#Function to convert 1s and 0s to [1, 0] and [0, 1]
	y = [[0, 1] if x[j][0] == 0 else [1, 0] for j in range(x.shape[0])]
	y = np.array(y)
	return y

def size(model): # Compute number of params in a model (the actual number of floats)
	return sum([sum(K.get_value(w).shape) for w in model.trainable_weights])

x = np.random.random((100000, 10))

#continuous classes: best results when using model1
y1 = np.array([[1/(1 + np.power(np.e, -np.sum(x[j])))] for j in range(x.shape[0])])
#descrete classes : best results when using model2
y2 = np.array([[int(np.sum(x[j])) % 2] for j in range(x.shape[0])])

#One output, sigmoid
model1 = Sequential()
model1.add(Dense(input_dim=10, output_dim=10))
model1.add(Dense(1))
model1.add(Dense(1))
model1.add(Activation('sigmoid'))

#2 outputs, softmax
model2 = Sequential()
model2.add(Dense(input_dim=10, output_dim=10))
model2.add(Dense(2))
model2.add(Activation('softmax'))


print size(model1), size(model2)

model1.compile(loss='mse', optimizer='sgd')
model1.fit(x, y2, validation_split=0.05, nb_epoch=20)

model2.compile(loss='mse', optimizer='sgd')
model2.fit(x, categorical(y2), validation_split=0.05, nb_epoch=20)
```

**Comment by Sandy4321** [comment]
May share py code file to try
